controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -154,6 +155,33 @@ func (r *BareMetalHostReconciler) Reconcile(request ctrl.Request) (result ctrl.R
 		}
 	}
 
+	// Consume inspect.metal3.io/hardwaredetails when either
+	// inspect.metal3.io=disabled or there are no existing HardwareDetails
+	if host.Status.HardwareDetails == nil || inspectionDisabled(host) {
+		objHardwareDetails, err := r.getHardwareDetailsFromAnnotation(host)
+		if err == nil && objHardwareDetails != nil {
+			reqLogger.Info("Setting HardwareDetails from annotation")
+			host.Status.HardwareDetails = objHardwareDetails
+			err = r.saveHostStatus(host)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err,
+					fmt.Sprintf("failed to save host status after adding HardwareDetails"))
+			}
+			return ctrl.Result{Requeue: true}, nil
+		}
+	} else {
+		// The hardwaredetails annotation is not needed, as the status is
+		// already present, so remove it.
+		if _, present := annotations[hardwareDetailsAnnotation]; present {
+			delete(annotations, hardwareDetailsAnnotation)
+			errStatus := r.Update(context.TODO(), host)
+			if errStatus != nil {
+				return ctrl.Result{}, errors.Wrap(errStatus, "Could not delete hardwaredetails annotation")
+			}
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
```

Comment:
The impression is that this new block, along with the previous one, represent an important preliminary step for properly restore a resource status from annotations, so they could be a good candidates to be refactored into a separate func - and this could also improve the readability of the reconcile loop.

In addition, what about coalescing the two update operations and postponing the `saveHostStatus` - if one of the two blocks modifies the status? In this way we could avoid one reconcile loop when both annotations are set

Reply:
I think it's unlikely a user will ever want to set both annotations, since the hardwaredetails annotation is a subset of the existing status annotation, also the status annotation will only ever be evaluated on the very first reconcile loop.

I can certainly refactor them into a separate function though, you're right that will make the reconcile function a bit more readable, and it's already pretty big.

Reply:
Ok I added a new function for the new code, if it's OK with you @andfasano I'll propose a follow-up for refactoring the existing code, since it's not strictly related to this feature PR?

Reply:
Sure. My feeling is that such kind of refactoring could also help in testing the feature (in more isolation)